proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,367 @@
+
+  [SE-0192]: https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md
+
+## Proposed solution
+
+The compiler will gain a new mode, called "library evolution mode". This mode will be off by default. A library compiled with this mode enabled is said to be _ABI-stable_.
+
+When a library is compiled with library evolution mode enabled, it is not an ABI-breaking change to modify the fields in a struct (to add, remove, or reorder them), and to add new enum cases (including with associated values). This implies that clients must manipulate fields and enum cases indirectly, via non-inlinable function calls. Information such as the size of the type, and the layout of its fields, becomes something that can only be determined at runtime. Types that reserve this flexibility are referred to as "resilient types".
+
+A new `@frozen` keyword will be introduced to allow library authors to opt out of this flexibility on a per-type basis. This keyword promises that stored instance properties within the struct will not be *added,* *removed,* or *reordered*, and that an enum will never *add,* *remove,* or *reorder* its cases (note removing, and sometimes reordering, cases can already be source breaking, not just ABI breaking). The compiler will use this for optimization purposes when compiling clients of the type. The precise set of allowed changes is defined below.
+
+This attribute has no effect when library evolution mode is off. In that case, the compiler acts as if *every* struct or enum in the library being compiled is "frozen".
+
+## Detailed design
+
+### Library evolution mode
+
+A new command-line argument, `-enable-library-evolution`, will enable this new mode.
+
+Turning on library evolution mode will have the following effects on source code:
+
+- The default behavior for enums will change to be non-frozen. This is the only change visible to _users_ of the library, who will need to use the `@unknown default:` technique described in SE-0192.
+- To ensure that all fields are initialized, inlinable `init`s of the type must delegate to a non-inlinable `init`.
+
+### `@frozen` on `struct` types
+
+When a library author is certain that there will never be a need to add fields to a struct in future, they may mark that type as `@frozen`. 
+
+This will allow the compiler to optimize away at compile time some calls that would otherwise need to be made at runtime (for example, it might access fields directly without the indirection).
+
+When compiling with binary stability mode on, a struct can be marked `@frozen` as long as it meets all of the following conditions:
+
+- The struct is *ABI-public* (see [SE-0193][]), i.e. `public` or marked `@usableFromInline`.
+- Every class, enum, struct, protocol, or typealias mentioned in the types of the struct's fields is ABI-public.
+- No fields have observing accessors (`willSet` or `didSet`).
```

Comment:
Maybe the forum is the best place to discuss this but I feel like this restriction is somewhat arbitrary and doesn’t buy us much.